### PR TITLE
[Mechanic] Shield/Hull Multipliers

### PR DESCRIPTION
**Pull request by Azure3141**
**Feature:** This PR implements the feature request detailed and discussed in issue #onceagainnooneaskedforthis

## Feature Details
This adds two new attributes, `shield multiplier` and `hull multiplier.` They do pretty much exactly what they say on the tin.

These differ from `shield protection` and `hull protection` in that they do not apply a multiplier to oncoming damage, instead affecting the ship's actual health points. This means shield generators and hull repair outfits do not have their production rates effectively multiplied.

## UI Screenshots
N/A. Really, what did you want to see here?

## Usage Examples
As an example, an outfit with ```"shield multiplier" 1``` would double the shields of the ship it was installed on (+100% shields, in line with the way the other multiplier attributes work). Obviously, this would be very powerful, so when implementing this on an outfit ~~like I have plans to do~~ I would recommend limiting the number that can be installed on each ship.

## Testing Done
I added each multiplier to my flagship and verified that the shields/hull were increased properly.

### Automated Tests Added
N/A

## Performance Impact
N/A

### CODE REVIEW

**Comment by bene-dictator** [comment]
Cause funny

**Comment by Saugia** [comment]
Working as intended.

**Comment by Ferociousfeind** [comment]
haunted

should probably close

**Comment by Quantumshark** [comment]
These need to be added to OutfitInfoDisplay.cpp in the same place as the shield generation and hull repair multipliers.

**Comment by petervdmeer** [comment]
The name doesn't seem to match the function; a "multiplier" suggests a multiplication; so a multiplication by 1 would be the standard/default value, while a value of 2 would double the amount of hull or shields.
The benefit of using a true multiplier is that it also can be used to reduce, for example a very nice weapon Outfit that works great, but that has a 0.5 shield-multiplier; effectively halving the amount of shields when the weapon is installed.

The "reducing" multiplier is also more likely to be used in the vanilla game. Outfits that add hull or shields are likely going to be plug-in-only (as mentioned in the relevant sections of https://github.com/endless-sky/endless-sky/wiki/CreatingOutfits).

File: source/Ship.cpp [hunk 1]
```diff
@@ -3324,13 +3325,25 @@ double Ship::Health() const
...
 	double minimumHull = MinimumHull();
 
 	return (hull > 0. ? minimumHull / hull : 0.);
 }
 
 
+// Get the maximum shield and hull values of the ship, accounting for multipliers.
+double Ship::MaxShields() const
+{
+	return attributes.Get("hull") * (1 + attributes.Get("shield multiplier"));
```
Comment (petervdmeer) [inline source/Ship.cpp:3338]: As mentioned in the comments, I prefer a multiplier to be a true multiplier.

**Review by petervdmeer — CHANGES_REQUESTED** [review]

File: source/Ship.cpp (hunk 1 above)
Comment (Zitchas) [reply source/Ship.cpp:3338]: I definitely agree with Petervdmeer on this. It is both easier to understand and simpler to have the multiplier be the actual multiplier.

**Comment by Amazinite** [comment]
The behavior of all existing multiplier attributes is to use `1. + attribute`. The reason it was done this way is because attributes default to 0, not 1. It's still a multiplier in the sense that you're adding or subtracting some _multiple_ of the base value.
https://github.com/endless-sky/endless-sky/blob/a282d003ae915a10fcf8db8f528c005d26bae37b/source/Ship.cpp#L2242-L2249

**Comment by Amazinite** [comment]
The above behavior also makes it so that adding two outfits with multipliers of -0.25 results in a total of -50%. If the multiplier were like how you're suggesting, two 0.75 outfits would add together to 1.5, meaning that two outfits that individually reduce your HP by 25% would somehow result in increasing it by 50%.

**Comment by petervdmeer** [comment]
@Amazinite Right; the current attributes system doesn't really support true multipliers and this pattern is already in use..

**Review by petervdmeer — APPROVED** [review]

**Comment by Ferociousfeind** [comment]
If we wanted a "true" multiplier, we could use exponents to display the internal, under-the-hood still-added-together attributes, instead of digging into the outfit adding code and handling weird exceptions to the simple addition rule.

The attribute defaults to 0, and some base raised to the 0th power is 1.0, multiplied by the base shield/hull values results in 100% shield/hull by default. The base we decide upon is the multiplier that +1 attribute translates into, while +2 translates into multiplying twice, and -3 translates into dividing three times.

The trick would be displaying the result of raising the base to the internal attribute value, rather than the attribute itself. Let's say shields and hull multipliers are equal to 2 to the power of the relevant attribute. If our shield booster gives +0.585 (or so), then it would display +50% (well, likely +50.0039...%), and the math would follow- four of these outfits would give a total of 2.34 attribute, and 2^2.34 gives you just about 5.0625x (more exactly, 5.06302...x) original shields amount, which is the expected amount- 1.5x1.5x1.5x1.5=5.0625.

Only issue is... exponentiation can be expensive, and probably shouldn't be performed every frame.

File: source/DamageProfile.cpp [hunk 2]
```diff
@@ -141,8 +141,10 @@ void DamageProfile::PopulateDamage(DamageDealt &damage, const Ship &ship) const
 			(1. + ship.DisruptionLevel() * .01);
 
 		damage.shieldDamage = (weapon.ShieldDamage()
-			+ weapon.RelativeShieldDamage() * attributes.Get("shields"))
+
```
Comment (quyykk) [inline source/DamageProfile.cpp:144]: Why the empty line :smile:

**Review by quyykk — COMMENTED** [review]
You also need to add tooltips

File: source/DamageProfile.cpp [hunk 3]
```diff
@@ -141,8 +141,9 @@ void DamageProfile::PopulateDamage(DamageDealt &damage, const Ship &ship) const
 			(1. + ship.DisruptionLevel() * .01);
 
 		damage.shieldDamage = (weapon.ShieldDamage()
-			+ weapon.RelativeShieldDamage() * attributes.Get("shields"))
+			+ weapon.RelativeShieldDamage() * ship.MaxShields())
 			* ScaleType(0., 0., attributes.Get("shield protection"));
+
```
Comment (Amazinite) [inline source/DamageProfile.cpp:146]:
```suggestion
```

File: source/PrintData.cpp [hunk 4]
```diff
@@ -164,7 +164,7 @@ namespace {
 				cout << ship.Cost() << ',';
 
 				auto mass = attributes.Mass() ? attributes.Mass() : 1.;
-				cout << attributes.Get("shields") << ',';
+				cout << ship.MaxShields() << ',';
 				cout << attributes.Get("hull") << ',';
```
Comment (Amazinite) [inline source/PrintData.cpp:168]: Why are we using MaxShields above this but not using MaxHull?
Line 165 of DamageProfile would also need this.

File: source/PrintData.cpp [hunk 5]
```diff
@@ -220,7 +220,7 @@ namespace {
...
```
Comment (Amazinite) [inline source/PrintData.cpp:224]: Ditto.

File: source/Ship.cpp [hunk 6]
```diff
@@ -3328,13 +3329,25 @@ double Ship::Health() const
...
```
Comment (Amazinite) [inline source/Ship.cpp:3342]:
```suggestion
	return attributes.Get("shields") * (1 + attributes.Get("shield multiplier"));
```

**Review by Amazinite — CHANGES_REQUESTED** [review]

File: source/Ship.cpp [hunk 7]
```diff
@@ -2478,13 +2478,25 @@ double Ship::Health() const
...
+{
+	return attributes.Get("shields") * (1 + attributes.Get("shield multiplier"));
+}
+
+
+double Ship::MaxHull() const
+{
+	return attributes.Get("hull") * (1 + attributes.Get("hull multiplier"));
+}
+
```
Comment (Amazinite) [inline source/Ship.cpp:2499]:
```suggestion


```

File: source/Ship.cpp [hunk 8]
```diff
@@ -2478,13 +2478,25 @@ double Ship::Health() const
 // Get the hull fraction at which this ship is disabled.
 double Ship::DisabledHull() const
 {
-	double hull = attributes.Get("hull");
+	double hull = MaxHull();
 	double minimumHull = MinimumHull();
 
 	return (hull > 0. ? minimumHull / hull : 0.);
 }
```
Comment (Amazinite) [inline source/Ship.cpp:2487]:
```suggestion


```

**Review by Amazinite — APPROVED** [review]

**Pull request merged**